### PR TITLE
bpo-47182: Fix crash by named unicode characters after interpreter reinitialization

### PR DESCRIPTION
**Pull request by tiran**
https://bugs.python.org/issue47182

Automerge-Triggered-By: GH:tiran

### CODE REVIEW

File: Objects/unicodeobject.c [hunk 1]
```diff
@@ -16085,6 +16085,8 @@ _PyUnicode_Fini(PyInterpreterState *interp)
     if (_Py_IsMainInterpreter(interp)) {
         // _PyUnicode_ClearInterned() must be called before _PyUnicode_Fini()
         assert(interned == NULL);
+
+        ucnhash_capi = NULL;
```
Comment (vstinner) [inline Objects/unicodeobject.c:16090]: Would you mind to add a short comment to explain why we do that? The code is fine, it's just for future readers.

**Review by vstinner — APPROVED** [review]
LGTM, but consider adding a comment ;-)

If you are motivated, you can add a test to test_embed using test_repeated_init_exec: see test_finalize_structseq() of test_embed for example. You can run arbitrary Python code, it's executed multiple times, each time the interpreter is created/destroyed in the same process.

**Comment by tiran** [comment]
I have added a comment and a test case.

File: Lib/test/test_embed.py [hunk 2]
```diff
@@ -343,6 +343,11 @@ def test_finalize_structseq(self):
         out, err = self.run_embedded_interpreter("test_repeated_init_exec", code)
         self.assertEqual(out, 'Tests passed\n' * INIT_LOOPS)
 
+    def test_ucnhash_capi_reset(self):
+        # bpo-47182: unicodeobject.c:ucnhash_capi was not reset on shutdown.
+        code = "print('\\N{digit nine}')"
+        out, err = self.run_embedded_interpreter("test_repeated_init_exec", code)
+        self.assertEqual(out, '9\n' * INIT_LOOPS)
```
Comment (vstinner) [inline Lib/test/test_embed.py:350]: Peeeerfect, thaaaanks!

**Review by vstinner — APPROVED** [review]
LGTM, thanks for adding a regression test!

**Comment by vstinner** [comment]
Are you sure that Python 3.10 and older are affected?

Clearing interned strings is new in Python 3.10. Python 3.11 changed that with immortal code objects and immortal strings.

In Python 3.9, strings were not really cleared by Py_Finalize(). They just leaked.

**Comment by tiran** [comment]
3.9 and earlier are definitely not affected. It doesn't hurt to backport the change to 3.10.

**Comment by vstinner** [comment]
> 3.9 and earlier are definitely not affected. It doesn't hurt to backport the change to 3.10.

Yeah, I agree to backport to 3.10. But you have to backport manually, _testembed doesn't have test_repeated_init_exec.

**Pull request merged**